### PR TITLE
teleport: init at v2.4.0

### PR DESCRIPTION
**Pull request by tomberek**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by tomberek** [comment]
Took a few tries to fix some whitespace conflicts to all-packages.nix, sorry.

File: pkgs/servers/teleport/default.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+in
+#stdenv.mkDerivation rec {
+buildGoPackage rec {
+  name = "teleport-${version}";
+  version = "2.4.0";
+  rev = "e9d66456309661c8da60e2cc66bc2a27d5739bef";
+
+  goPackagePath = "github.com/gravitational/teleport";
+  buildPhase = ''
+runHook preBuild
```
Comment (nlewo) [inline pkgs/servers/teleport/default.nix]: Indentation

**Review by nlewo — CHANGES_REQUESTED** [review]
I didn't really reviewed your PR yet since I think you should try to use go2nix V2 branch. This version of go2nix uses Gopkg.lock file which is available in teleport, so you should be able to generate `deps.nix` file.

**Comment by tomberek** [comment]
Thanks. I was not aware of the V2. This application does have a few quirks to build/install. I’ll push an update if the end result is cleaner.

**Comment by tomberek** [comment]
@nlewo Took a while to get v2 working. The deps.nix file is now populated. Sadly much of the cruft was still needed to build correctly.

**Comment by nlewo** [comment]
@tomberek I agree it's not documented... but there is a `default.nix` inside :)

Did you try to specify which subpackages you want to build? For instance:
```
buildGoPackage rec {
  name = "teleport-unstable-${version}";
  version = "2.4.0";

  goPackagePath = "github.com/gravitational/teleport";
  subPackages = [ "tool/tctl" ];
  # This repo has a private submodule "e" which fetchgit cannot handle
  # without failing. Using fetchurl instead on the latest tagged release.
  src = fetchurl {
    url = "https://github.com/gravitational/teleport/archive/v2.4.0.tar.gz";
    sha256 = "01qcr6vml3ias3gbkqjnb5fmc6ap1fd5mi52ygn3agafy63jb5rd";
  };
  goDeps = ./deps.nix;

  meta = {
    description = "A SSH CA management suite";
    homepage = "https://gravitational.com/teleport/";
    license = stdenv.lib.licenses.asl20;
    maintainers = [ stdenv.lib.maintainers.tomberek ];
    platforms = stdenv.lib.platforms.linux;
  };
}
```

**Comment by tomberek** [comment]
@nlewo Thanks for the simplifications. (this is my first go package for nixpkgs).

File: pkgs/servers/teleport/default.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    cat build/webassets.zip >> $NIX_BUILD_TOP/go/bin/teleport
+    rm -fr build/webassets.zip
+    cd $NIX_BUILD_TOP/go/bin
+    zip -q -A teleport
+    popd
+    '';
+
+  dontStrip = true;
+
+  # This repo has a private submodule "e" which fetchgit cannot handle
```
Comment (nlewo) [inline pkgs/servers/teleport/default.nix]: It seems the commit you were using (`df6caeae7d90e520478244d583cd9e60d924fb32`) is not the commit ID for the tag 2.4.0. So, maybe there is a diff between the deps you genreated and the soruce code you are using.
- You can fetch an archive for a particular commit ID on github or
- Set this attribute: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgit/default.nix#L16 or
- Generate deps by checking out tag 2.4.0

File: pkgs/servers/teleport/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# This file was generated by https://github.com/kamilchm/go2nix v2.0-dev
+{ stdenv, buildGoPackage, zip, fetchurl, fetchgit, fetchhg, fetchbzr, fetchsvn }:
+
+buildGoPackage rec {
+  name = "teleport-unstable-${version}";
```
Comment (nlewo) [inline pkgs/servers/teleport/default.nix]: Remove "unstable" from the name

**Review by nlewo — CHANGES_REQUESTED** [review]
Last remarks form my side :)

File: pkgs/servers/teleport/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  /*
+  src = fetchgit {
+    rev = "df6caeae7d90e520478244d583cd9e60d924fb32";
+    url = "https://github.com/gravitational/teleport";
+    sha256 = "05f0a0z831rdn77ji5p6xd27cpyj2c2jncivaby5qkqj1w0kdpn5";
+  };
+  */
+
+  goDeps = ./deps.nix;
```
Comment (kamilchm) [inline pkgs/servers/teleport/default.nix]: Try to ignore `goDeps` here, there's missing vendor scaning logic in go2nix v2. Packages from `Gopkg.lock` should be included only when they doesn't exists in `vendor` dir.

Comment (nlewo) [reply pkgs/servers/teleport/default.nix]: I confirm this works well without `deps.nix`.
I didn't realize `goBuildPackage` uses the `vendor` directory and I don't really know what to think about this: I like to be able to know what are the build requires by querying the Nix store and this is no longer possible when the `vendor` directorys used. But this is another story:)

For now, I think we could remove `deps.nix`.

File: pkgs/servers/teleport/default.nix (hunk 3 above)
Comment (tomberek) [reply pkgs/servers/teleport/default.nix]: done

File: pkgs/servers/teleport/default.nix (hunk 2 above)
Comment (tomberek) [reply pkgs/servers/teleport/default.nix]: Correct, that was master a few days ago. Good catch. But we're not longer using deps... so meh.

File: pkgs/servers/teleport/default.nix (hunk 4 above)
Comment (tomberek) [reply pkgs/servers/teleport/default.nix]: removing....

**Comment by tomberek** [comment]
@kamilchm @nlewo Thanks for the review!

**Review by nlewo — APPROVED** [review]

**Review by kamilchm — APPROVED** [review]

**Comment by nlewo** [comment]
@Mic92 could you have a look please?

File: pkgs/servers/teleport/default.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    popd
+    '';
+
+  dontStrip = true;
+
+  meta = {
+    description = "A SSH CA management suite";
+    homepage = "https://gravitational.com/teleport/";
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ stdenv.lib.maintainers.tomberek ];
```
Comment (Mic92) [inline pkgs/servers/teleport/default.nix]: @kimburgess do you also want to maintain this package?

Comment (KimBurgess) [reply pkgs/servers/teleport/default.nix]: I'm all good staying out of this one - I'm not affiliated with Teleport, my other PR was just to get it into the repo for ease of use.

If there's an update that needs repacking in the future I'm happy to step up as a contributor / maintainer then and help out.

**Pull request merged**